src/ray/raylet/worker_pool.cc

```diff
@@ -725,29 +724,37 @@ void WorkerPool::TryKillingIdleWorkers() {
       auto rpc_client = worker->rpc_client();
       RAY_CHECK(rpc_client);
       rpc::ExitRequest request;
-      rpc_client->Exit(
-          request, [this, worker](const ray::Status &status, const rpc::ExitReply &r) {
-            if (!status.ok()) {
-              RAY_LOG(ERROR) << "Failed to send exit request: " << status.ToString();
-              return;
-            }
+      rpc_client->Exit(request, [this, worker](const ray::Status &status,
+                                               const rpc::ExitReply &r) {
+        if (!status.ok()) {
+          RAY_LOG(ERROR) << "Failed to send exit request: " << status.ToString();
+          return;
+        }
 
-            if (r.success()) {
-              auto &worker_state = GetStateForLanguage(worker->GetLanguage());
-              // If we could kill the worker properly, we remove them from the idle pool.
-              if (RemoveWorker(worker_state.idle, worker)) {
-                // If the worker is not idle at this moment, we don't mark them dead.
-                // In this case, the core worker will exit the process after
-                // finishing the assigned task, and DisconnectWorker will handle this
-                // part.
-                if (!worker->IsDead()) {
-                  worker->MarkDead();
-                }
-              }
+        if (r.success()) {
+          auto &worker_state = GetStateForLanguage(worker->GetLanguage());
+          // If we could kill the worker properly, we remove them from the idle pool.
+          if (RemoveWorker(worker_state.idle, worker)) {
+            // If the worker is not idle at this moment, we don't mark them dead.
+            // In this case, the core worker will exit the process after
+            // finishing the assigned task, and DisconnectWorker will handle this
+            // part.
+            if (!worker->IsDead()) {
+              worker->MarkDead();
             }
-            RAY_CHECK(pending_exit_idle_workers_.count(worker->WorkerId()));
-            RAY_CHECK(pending_exit_idle_workers_.erase(worker->WorkerId()));
-          });
+          }
+        } else {
+          // We re-insert the idle worker to the back of the queue if it fails to kill the
+          // worker (e.g., when the worker owns the object). Without this, if the first N
+          // workers own objects, it can't kill idle workers that are >= N+1.
+          const auto &idle_pair = idle_of_all_languages_.front();
```

Comment:
Why do we always pop the first idle worker here?

Reply:
This part will just reorder the FIFO queue.